### ISSUE TITLE
508-defect-2 [SCREENREADER]: Error messages must be announced when tabbing through fields

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#22298]
# [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** Josh

## User Story or Problem Statement

As a screen reader user, I want to know what is wrong with a field when I tab into it in forms mode when an error exists.

## Details

Errors are currently _not_ announced to screen reader users while in forms mode (navigating by `tab`). This is because they aren't programmatically linked to their respective fields (either by including them in the `<label>` or `<legend>` or using `aria-describedby`. As a result, should a user return back to an error field earlier in the form OR be redirected there upon attempting to continue, they _won't_ be able to hear the error message again unless they switch out of forms mode.

## Acceptance Criteria

Example:
- [ ] Upon tabbing into a field with an error, the error is read out to the user

## Steps to Recreate

1. Enter `https://staging.va.gov/health-care/covid-19-vaccine/stay-informed/form#content` in browser
2. Start screenreading device
3. Create errors in all fields (e.g. enter abc into phone)
4. Attempt to continue
5. Tab through all fields again and confirm none of the errors are read out 

#### Current

```html
<div class="schemaform-field-template usa-input-error ">
  <!-- Label -->
  <label id="root_email-label" class="usa-input-error-label schemaform-label" for="root_email">
    Email address<span class="schemaform-required-span">(*Required)</span>
  </label>
  <!-- Error message -->
  <span class="usa-input-error-message" role="alert" id="root_email-error-message">
    <span class="sr-only">Error</span> 
      Please enter a valid email address
  </span>
  <!-- Field -->
  <div class="schemaform-widget-wrapper">
    <input type="email" id="root_email" name="root_email" value="pdf" >
  </div>
<div>
```

#### Suggested

```html
<div class="schemaform-field-template usa-input-error ">
  <!-- Label -->
  <label id="root_email-label" class="usa-input-error-label schemaform-label" for="root_email">
    Email address<span class="schemaform-required-span">(*Required)</span>
    <!-- Error message is now within the label -->
    <span class="usa-input-error-message" role="alert" id="root_email-error-message">
    <span class="sr-only">Error</span> 
      Please enter a valid email address
    </span>
  </label>
  <!-- Field now has aria-invalid="true" -->
  <div class="schemaform-widget-wrapper">
    <input aria-invalid="true" type="email" id="root_email" name="root_email" value="pdf">
  </div>
<div>
```

#### Current

```html
<div class="schemaform-field-template usa-input-error ">
  <!-- Label -->
  <label id="root_email-label" class="usa-input-error-label schemaform-label" for="root_email">
    Email address<span class="schemaform-required-span">(*Required)</span>
  </label>
  <!-- Error message -->
  <span class="usa-input-error-message" role="alert" id="root_email-error-message">
    <span class="sr-only">Error</span> 
      Please enter a valid email address
  </span>
  <!-- Field -->
  <div class="schemaform-widget-wrapper">
    <input type="email" id="root_email" name="root_email" value="pdf" >
  </div>
<div>
```

#### Suggested

```html
<div class="schemaform-field-template usa-input-error ">
  <!-- Label -->
  <label id="root_email-label" class="usa-input-error-label schemaform-label" for="root_email">
    Email address<span class="schemaform-required-span">(*Required)</span>
  </label>
  <!-- Error message -->
  <span class="usa-input-error-message" role="alert" id="root_email-error-message">
    <span class="sr-only">Error</span> 
      Please enter a valid email address
  </span>
  <!-- Field is now linked to the error with aria-describedby -->
  <div class="schemaform-widget-wrapper">
    <input aria-describedby="root_email-error-message" type="email" id="root_email" name="root_email" value="pdf" >
  </div>
<div>
```

## Screenshots or Trace Logs

https://user-images.githubusercontent.com/14154792/112867965-83bc3100-9089-11eb-866f-7c313a02b93c.mov

<img width="1203" alt="Screen Shot 2021-03-29 at 11 56 42 AM" src="https://user-images.githubusercontent.com/14154792/112867981-8880e500-9089-11eb-8fed-5b29e03c7125.png">

**Comment by cohnjesse** [comment]
@joshkimux I totally agree that this needs to be fixed however this is a form system issue and therefore should not be a launch blocker for anyone. This should be fixed at the platform level and should not block anyone from launching before it is fixed there, fixing it at the app level only creates further complications in the codebase that will need to be cleaned up by the platform team later. @batemapf @artsymartha68 for visibility.

**Comment by joshkimux** [comment]
@cohnjesse Thanks for the context, I'll bump this down to a level 2 to make way for launch. Do we have a sitewide ticket made for this?

**Comment by 1Copenut** [comment]
@joshkimux We don't have a ticket for this fix specifically. I did create one for handling IE11 + JAWS reading errors twice: https://github.com/department-of-veterans-affairs/va.gov-team/issues/314

There's a question of how much runway IE11 still has, but that's outside the scope of your question. Please go ahead and open a new ticket and tag it with `vsp-design-system-team` and the regular labels for this work. Please also link it to #314 above. Thank you!

**Comment by rsmithadhoc** [comment]
There have been many changes to form elements in the design system, plus the form library (which we contributed to). In both the standalone components and the special versions for the form library, they are correctly announcing errors. I will close this out.
